### PR TITLE
fix: allow catalog groups with no id and fix validation of links and reference matching

### PR DESCRIPTION
**Pull request by fsuits**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [x] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary
Catalog group id's are optional so now create ids for any groups missing id's on import and give warning
Fixed bug counting links and references during validation
Added new load_validate.py file to validate files during load - which will also create group id's if needed
Then replaced all functions like load_top_level... to use the load_validate instead.

## Key links:

- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

**Review by enikonovad — COMMENTED** [review]
I was playing around and got a bit confused with the use case. Say we have an existing catalog `ABC_catalog` that contains something like this (notice no group id):
```
"groups": [
      {
        "title": "",
        "controls": [
          {
            "id": "abc-1",
            "title": "ABC Test",
            "parts": [
              {
                "id": "abc-1_smt",
                "name": "statement",
                "prose": "Some Prose."
              }
            ]
          }
        ]
      }
   ]
```
And it is markdown representation.
Now we run `trestle author catalog-assemble -v --set-parameters --markdown md_catalogs/ABC_catalog --output ABC_catalog` for this catalog. 
It will result in error since control ids do not match the group.

My question is, is this a possible scenario? How can we expect the control ids to match group id?

**Comment by fsuits** [comment]
OK I had to make many changes, but to summarize I rely on validation to create group id's as needed - but some of the load functions were not doing validation.  I added a new file load_validate to do loads and validations without causing circular imports.  I then changed all functions like load_top_level_model to use these methods - but they are still present in the tests, but that should be ok.

**Review by enikonovad — APPROVED** [review]
LGTM - looks solid, one suggestion is to add a component test that would import a catalog with no id, generate markdown and assemble back. And test that group id is assigned on all stages. Otherwise looks good.


```

trestle import -f <path_to_catalog>/CATALOG_WITH_NO_GROUP_ID.json -o A_CATALOG
trestle author catalog-generate --name A_CATALOG --output md_A_CATALOG
trestle author catalog-assemble --markdown md_A_CATALOG --output A_CATALOG

```

**Comment by fsuits** [comment]
Your additional request ended up being important.  It turns out that Import runs the validation *after* importing the file, which means the validation doesnt get to fix the file.  I needed to add an early validation of the model in memory immediately after loading it.  This will prevent saving the model if it is not valid - but it will allow fixing it in ways that trestle is able to do during validation.  Validation is then run again on the saved file - as a final check.  A more elaborate test has been added to do the full import, generate markdown, then assemble with a catalog missing group id.

File: trestle/core/commands/import_.py [hunk 1]
```diff
@@ -66,6 +68,14 @@ def _run(self, args: argparse.Namespace) -> int:
 
             model_read, parent_alias = fetcher.get_oscal(True)
 
+            # validate the loaded model in memory before writing out
+            # this will do any needed fixes to the file, such as assign missing catalog group ids
+            args_validate = argparse.Namespace(mode=const.VAL_MODE_ALL)
+            validator: Validator = validator_factory.get(args_validate)
+            if not validator.model_is_valid(model_read, True):
+                logger.warning(f'Validation of file to be imported {input_uri} did not pass.  Import failed.')
+                return CmdReturnCodes.COMMAND_ERROR.value
```
Comment (enikonovad) [inline trestle/core/commands/import_.py:77]: Since this ends the execution it should be an error instead.

**Review by enikonovad — APPROVED** [review]
LGTM - please use error instead of warning to confirm to logging in the rest of the commands

**Pull request merged**